theano/gpuarray/c_code/dnn_fwd.c

```diff
@@ -167,31 +204,17 @@ APPLY_SPECIFIC(conv_fwd)(PyGpuArrayObject *input, PyGpuArrayObject *kerns,
             cuda_exit(c->ctx);
             return 1;
           }
+          prev_algo.algo = algo;
+          // no tensor_op returned from Get()
+          prev_algo.mathType = mathtype = CUDNN_DEFAULT_MATH;
+	  // fprintf(stderr, "(cudnnGetConvolutionForwardAlgorithm: (err:%d), algo: %d\n", err, algo);
       }
-      prev_algo = algo;
-    } else {
-      algo = prev_algo;
+    } else { 
+      algo = (cudnnConvolutionFwdAlgo_t)prev_algo.algo;
+      worksize = prev_algo.wsSize;
+      mathtype = prev_algo.mathType;
     }
-
-    #ifdef DEBUG
-    if (0 != theano_enum_to_string_cudnnConvolutionFwdAlgo_t(algo, algorithm_name))
-        return 1;
-    // NB: This is printed only when algorithm is chosen at runtime.
-    if (reuse_algo)
-        fprintf(stderr, "(reused %s)\n", algorithm_name);
-    else
-        fprintf(stderr, "(using %s)\n", algorithm_name);
-    #endif
-
-    if (params->choose_once) {
-      reuse_algo = 1;
-    } else {
-      for (unsigned int i = 0; i < PyGpuArray_NDIM(input); i++) {
-        prev_img_dims[i] = PyGpuArray_DIM(input, i);
-        prev_kern_dims[i] = PyGpuArray_DIM(kerns, i);
-      }
-    }
-  }
+  } else { /* choose_algo */
```

Comment:
Currently there are fallbacks for fft/fft_tiling and a fallback for 3D supported conv algos in dnn_fwd.
1. We need to decide if we put a fallback for 3D supported conv algos in all computations (ie. in dnn_gw and dnn_gi too)
2. Checkings for fft/fft_tiling seems quite old, so we may need to update them. Or, as related cases should raise a CUDNN_STATUS_NOT_SUPPORTED, we may drop these checkings and let worksize fallback do the job.

I think a discussion is needed for that, @nouiz @abergeron . But in any case, we need an entry for fallback checkings!

I am preparing new commits about that. I can push them in next minutes, so that we can discuss about it tonight and tomorrow.